docs/design/20190914.certificaterenewal.md

```diff
@@ -0,0 +1,170 @@
+
+Each volume directory is named using the format of:
+
+```
+cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID
+```
+
+This naming scheme ensures full namespacing of directory names whereby even if a
+single pod contains multiple cert-manager-csi volume mounts, each volume ID is
+ensured to be unique in this location. If anything external has created a
+directory name and causes a conflict, the driver should fail to create the
+volume until it has been resolved.
+
+Inside the pod directory contains the metadata file of the volume as well as the
+mount directory. Although the mount directory can have any child directory
+structure since a user can specify the key and certificate location, typically
+the directory will only contain 2 files at its root containing the key and
+certificate.
+
+```
+/cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID/
+-- /metadata.json
+-- /data/
+---- /crt.pem
+---- /key.pem
+```
+
+### Meta Data File
+
+In order to retain information about the volume attributes, even if the driver
+has been restarted, data needs to be persisted on local storage. If this method
+was not taken, we would require to query the API server for all pods in the
+cluster, then sort the pods that have a cert-manager-csi volume attached and
+match by node. By keeping a local store, we eliminate our need for API server
+accesses to pods and reduce complexity. Volume mount attributes are immutable so
+the data shall not change during a pods life cycle.
+
+The contents of the meta data includes the volume attributes
+`map[string]string`, the volume ID and target path. With this a full recreation
+of the information for each volume can be made.
+
+The file will be stored as a `.json` file type. This is a widely used and
+understood file type and Go has good support for converting to and from Go
+structs and Json string.
+
+### Certificate Watchers
+
+Each volume on the driver node will have its own long running Go routine that
+waits for the certificate to be ready for renewal. The default renew before
+duration will be the same used as in the cert-manager API (30 days). This
+option can be changed using a volume attribute
+`csi.certmanager.k8s.io/renew-before` that takes a Go time duration string.
+
+Once the Go routine has reached this duration, the certificate should be marked
+for renewal and trigger a re-fetch of a certificate.
+
+If the volume is unpublished (pod deleted), then the corresponding Go routine
+should also be killed.
+
+The Go routine can be prevented, and as such renewal not to take place, via the
+volume attribute `csi.certmanager.k8s.io/disbale-auto-renew` set to `true`.
+
+### Renewal Procedure
+
+Once a certificate has been marked for renewal, the driver will attempt to
+discover if an existing cert-manger `CertificateRequest` exists. If so, the
+driver should delete it. The driver should then create a new private key, only
+if the volume attribute `csi.certmanager.k8s.io/reuse-private-key` is not set
```

Comment:
"is not set to `true`" forced my mind to bend a bit.. can we rephrase this? Also, what's the default behaviour here? Generate a new private key?

Reply:
Indeed yes. You think it should be the reverse?